### PR TITLE
added event grid send message code

### PR DESCRIPTION
**Pull request by bartr**
added a Go app and shell script to send a message to event grid

all new files so no merge conflicts

### CODE REVIEW

File: sendmessage/main.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	// execute and get a response
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
```
Comment (day-jeff) [inline sendmessage/main.go]: Potentially more concise error handling code...

if resp, err := client.Do(req); err != nil {
	log.Println(err)
	return
}

File: sendmessage/main.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+	// read the response
+	_, err = ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
```
Comment (day-jeff) [inline sendmessage/main.go]: Overkill? If .Do was successful, won't ReadAll always be too?

File: sendmessage/main.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+var key *string
+var fname *string
+var lname *string
+
+func main() {
+	endpoint = flag.String("endpoint", "", "Event Grid Endpoint")
+	key = flag.String("key", "", "Event Grid Key")
+	fname = flag.String("fname", "John", "First Name")
+	lname = flag.String("lname", "Doe", "Last Name")
+	flag.Parse()
```
Comment (day-jeff) [inline sendmessage/main.go]: Not sure your intended use for this utility, but do you really want John Doe to be used as the default value?

File: sendmessage/main.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	// read the response
+	_, err = ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	// close the request
+	resp.Body.Close()
```
Comment (day-jeff) [inline sendmessage/main.go]: Better to group Close() .Do() call:
 
     defer resp.Body.Close()

File: sendmessage/sendmsg.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
+# requires Azure cli and access to the subscription
+# run using source or . or it won't work because of the return
+
+
+if [[ -z ${m4key} ]] || [[ -z ${m4endpoint} ]]
+then
+        if [[ -z $1  ]]
+        then
+                echo "usage: $0 subscriptionName"
```
Comment (day-jeff) [inline sendmessage/sendmsg.sh]: This usage summary isn't technically correct, as you can optionally pass in the end point and key.

File: sendmessage/sendmsg.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+                exit 1
+        fi
+
+        az account set -s $1
+        m4endpoint=$(az eventgrid topic show --resource-group 7-11-Dev --name minus4dev --query endpoint --output tsv)
+        m4key=$(az eventgrid topic key list --resource-group 7-11-Dev --name minus4dev  --output tsv --query key1)
+fi
```
Comment (day-jeff) [inline sendmessage/sendmsg.sh]: Minor nit: If someone passes is one value but not the other, the value that was passed in is overridden.

File: sendmessage/main.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	wrapper = append(wrapper, genMessage())
+
+	// Marshal the json
+	json, err := json.Marshal(wrapper)
+
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (bartr) [reply sendmessage/main.go]: It *shouldn't* happen ... but, generally, if a function returns an err, I check it.

File: sendmessage/main.go (hunk 3 above)
Comment (bartr) [reply sendmessage/main.go]: It's a quick test to verify end to end functionality. 

we could change the flag to not have a default which would mean you would have to include on the command line every time.

I'm OK either way.

File: sendmessage/main.go (hunk 4 above)
Comment (bartr) [reply sendmessage/main.go]: that's more idiomatic. we should make that change.

File: sendmessage/sendmsg.sh (hunk 5 above)
Comment (bartr) [reply sendmessage/sendmsg.sh]: you can't pass in the endpoint and the key. they are read from environment variables which are set when you "source" the script the first time.

The script runs a lot faster without having all of the az calls, so if you want to test sending 5 messages, it's faster.

File: sendmessage/sendmsg.sh (hunk 6 above)
Comment (bartr) [reply sendmessage/sendmsg.sh]: you can only pass in the subscription name. if one of the environment variables is not set, both get set.

File: sendmessage/main.go (hunk 2 above)
Comment (bartr) [reply sendmessage/main.go]: Potentially not. The response could overflow memory or the socket could get abruptly closed are two things I can think of.

File: sendmessage/main.go (hunk 1 above)
Comment (bartr) [reply sendmessage/main.go]: I'm OK with either approach, but we should be consistent. I see it both ways in Go code.

I would say if it's one statement checking the error, we should combine them like above. If it's more than one statement or not checking err, we should keep them on separate lines.

Feedback?

File: sendmessage/main.go (hunk 4 above)
Comment (bartr) [reply sendmessage/main.go]: I made that change and checked it in

**Comment by day-jeff** [comment]
I say merge it

**Pull request merged**